folio-system-user/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,157 @@
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class SystemUserService {
+
+  private final ExecutionContextBuilder contextBuilder;
+  private final SystemUserProperties systemUserProperties;
+  private final FolioEnvironment environment;
+  private final AuthnClient authnClient;
+  private final PrepareSystemUserService prepareUserService;
+  private Cache<String, SystemUser> systemUserCache;
+
+  /**
+   * Get authenticate system user.
+   * Get from cache if present (or getSystemUser otherwise). Call login expiry endpoint in case
+   * access token expired. Call login endpoint in case login-expiry endpoint returns null or it
+   * doesn't exist
+   *
+   * @param tenantId The tenant name
+   * @return {@link SystemUser} with token value
+   */
+  public SystemUser getAuthedSystemUser(String tenantId) {
+    if (systemUserCache == null) {
+      return getSystemUser(tenantId);
+    }
+
+    var user = systemUserCache.get(tenantId, this::getSystemUser);
+    var userToken = user.token();
+    var now = Instant.now();
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      return user;
+    }
+
+    systemUserCache.invalidate(tenantId);
+    user = getSystemUser(tenantId);
+    systemUserCache.put(tenantId, user);
+
+    return user;
+  }
+
+  /**
+   * Authenticate system user and return token value.
+   *
+   * @param user {@link SystemUser} to log with
+   * @return token value
+   */
+  public UserToken authSystemUser(SystemUser user) {
+    var token = getTokenWithExpiry(user);
+    if (token == null) {
+      token  = getTokenLegacy(user);
+    }
+    return token;
+  }
+
+  @Autowired(required = false)
+  public void setSystemUserCache(Cache<String, SystemUser> systemUserCache) {
+    this.systemUserCache = systemUserCache;
+  }
+
+  private SystemUser getSystemUser(String tenantId) {
+    log.info("Attempting to issue token for system user [tenantId={}]", tenantId);
+    var systemUser = SystemUser.builder()
+        .tenantId(tenantId)
+        .username(systemUserProperties.username())
+        .okapiUrl(environment.getOkapiUrl())
+        .build();
+
+    // create context for authentication
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var token = authSystemUser(systemUser);
+      systemUser = systemUser.withToken(token);
+      log.info("Token for system user has been issued [tenantId={}]", tenantId);
+    }
+    // create context for user with token for getting user id
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var userId = prepareUserService.getFolioUser(systemUserProperties.username())
+          .map(UsersClient.User::id).orElse(null);
+      return systemUser.withUserId(userId);
+    }
+  }
+
+  private UserToken getTokenLegacy(SystemUser user) {
+    var responseOptional =
+      ofNullable(authnClient.login(new UserCredentials(user.username(), systemUserProperties.password())));
```

Comment:
`authnClient.login` is a Feign client method and returns `ResponseEntity<>`
what is the case when this method could return null?

Reply:
According to Sonar the response could be null. I was a bit surprised by this as well.

Reply:
Please review the change